boot/zephyr/main.c

```diff
@@ -101,6 +107,29 @@ void main(void)
             ;
     }
 
+#ifdef CONFIG_MCUBOOT_SERIAL
+
+    struct device *detect_port;
+    u32_t detect_value;
+
+    detect_port = device_get_binding(CONFIG_GPIO_NRF5_P0_DEV_NAME);
```

Comment:
I think it's OK to start out the merge like this since the initial users are all on nRF devices, but there is a better way to do this than putting SoC specific CONFIG option names in the main sources of the file which I think should be fixed in the future.

The basic idea is that, instead of putting CONFIG_GPIO_NRF5_P0_DEV_NAME in the source itself, you should pick a name that is mcuboot-specific and use it here, like this:

```
#define SERIAL_DETECT_GPIO_DEVICE "mcuboot-serial-gpio"

...

detect_port = device_get_binding(SERIAL_DETECT_GPIO_DEVICE);
```

Then, you can use SoC or device-specific .conf file fragments to set CONFIG_GPIO_NRF5_P0_DEV_NAME to "mcuboot-serial-gpio".

Here is an example:

https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/drivers/led_ws2812/src/main.c

Note how SPI_DEV_NAME is just "ws2812_spi". Then, board-specific .conf file fragments configure the appropriate device:

https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/drivers/led_ws2812/boards/96b_carbon.conf

To eliminate duplication, these fragments are merged into the final configuration along with the top-level prj.conf:

https://github.com/zephyrproject-rtos/zephyr/blob/master/samples/drivers/led_ws2812/CMakeLists.txt#L1

(Perhaps DT could be used as well, but I don't know how.)

Reply:
@mbolivar  - I agree - lest merge this like is now and rework it later - thank for tutorial.

Reply:
@nvlsianpu :+1: np

Reply:
So looking at this series again and some of the discussion, I think it's almost there, strangely. This series adds BOOT_SERIAL_DETECT_PORT in the Kconfig, sets it to CONFIG_GPIO_NRF5_P0_DEV_NAME, but then uses the nRF value here instead of the Kconfig value... why? 

Anyway, the cleanup commit should just use CONFIG_BOOT_SERIAL_DETECT_PORT here. You can add board-specific config fragments as described in my earlier comments to replace the "default foo if BOARD_NRFxxx" entries for BOOT_SERIAL_DETECT_PIN. Other changes can be made at SoC or board level with some mix of Kconfig edits, DT changes, and board config fragments.